Demo drivers attributes

ranechita opened the pull request:
Added attributes for the virtual adc&dac.

amiclaus reviewed (commented):
Add a PR description please. Not mandatory, but useful to summarize the changes/additions from the commits.

Also, try to resolve in the first place the file conflicts detected on the branch.

amiclaus commented on drivers/dac/dac_demo/dac_demo.c:
```diff
@@ -85,6 +92,128 @@ int32_t dac_demo_remove(struct dac_demo_desc *desc)
 }
 
+
+int32_t dac_get_number_of_channels(void* dev)
```
Maybe:

`int32_t dac_get_number_of_channels(void* dev, uint32_t *ch_nr)`
`{`
` struct dac_demo_desc *desc = dev;`
` *ch_nr = dev->active_ch;`
` return SUCCESS;`
`}`

In this manner you will keep the same format as other functions, returning an error code.

ranechita replied:
removed the function completely as it was no longer used

amiclaus replied:
same mentions for `int32_t adc_get_number_of_channels(void* dev)`
https://github.com/analogdevicesinc/no-OS/pull/935/files#diff-de54365b8e51917ba102b742e5f00709bd9026a55d14db2e564c12dd67384d31R115

amiclaus commented on drivers/dac/dac_demo/dac_demo.c:
```diff
@@ -85,6 +92,128 @@ int32_t dac_demo_remove(struct dac_demo_desc *desc)
+	return 1;
+}
+
+int32_t dac_write_samples(void* dev, uint16_t* buff, uint32_t samples)
```
Add function header comment.

Also, maybe return an error code instead or `samples`.

Otherwise, you should switch the return type for the function to `uint32_t` since the number of samples cannot be negative.

ranechita replied:
The function should return the number of samples because they are used, somewhere else, but I changed it to uin32_t

amiclaus commented on projects/iio_demo/src.mk:
```diff
@@ -8,13 +8,25 @@ SRCS +=	$(NO-OS)/util/xml.c						\
 	$(INCLUDE)/list.h						\
 	$(INCLUDE)/util.h						\
 	$(INCLUDE)/error.h
 
```
nitpick: extra line

amiclaus commented on projects/iio_demo/src/app/main.c:
```diff
@@ -194,12 +213,56 @@ int main(void)
+		.dev_ch_attr = 3311
+	};
+	status = adc_demo_init(&adc_desc, &adc_init_par);
+	if (status != 0)
```
try to use error codes instead of raw values. For example `status != SUCCESS`. Check through the entire code.

amiclaus commented on projects/iio_demo/src/app/main.c:
```diff
@@ -46,6 +46,11 @@
+#include "iio_adc_demo.h"
+#include "dac_demo.h"
+#include "iio_dac_demo.h"
+
```
nitpick: extra line

amiclaus commented on drivers/dac/dac_demo/iio_dac_demo.c:
```diff
@@ -0,0 +1,95 @@
+		};
+		iio_dac_channels[i] = ch;
+	}
+
```
nitpick: extra line

amiclaus commented on drivers/adc/adc_demo/iio_adc_demo.c:
```diff
@@ -0,0 +1,100 @@
+	return SUCCESS;
+}
+
+
```
nitpick: extra lines

amiclaus commented on drivers/adc/adc_demo/adc_demo.c:
```diff
@@ -85,6 +112,129 @@ int32_t adc_demo_remove(struct adc_demo_desc *desc)
+ * @param samples - number of samples to receive
+ * @return SUCCESS in case of success, negative error code otherwise.
+*******************************************************************************/
+int32_t adc_read_samples(void* dev, uint16_t* buff, uint32_t samples)
```
Maybe return an error code instead or `samples`.

Otherwise, you should switch the return type for the function to `uint32_t` since the number of samples cannot be negative.

Also, the header comment @return does not match the actual implementation.

ranechita commented:
v3

- removed extra lines
- updated @return inheader comments of adc_read_samples to match the actual implementation
- added `if(status != SUCCESS)` instead of `if(status != 0)` in main
-  added comment for the function `dac_write_samples`
- deleted the function `dac_get_number_of_channels`

adrimbarean commented on drivers/adc/adc_demo/adc_demo.c:
```diff
@@ -85,6 +110,81 @@ int32_t adc_demo_remove(struct adc_demo_desc *desc)
+	uint32_t ch = -1;
+
+	if(desc->loopback == NULL)
+	{
```
nitpick: opening braces on the same line as the 'if' statement. [Coding style document](https://www.kernel.org/doc/html/v4.10/process/coding-style.html) will provide more information.

mirceacaprioru commented on drivers/adc/adc_demo/adc_demo.c:
```diff
@@ -85,6 +110,81 @@ int32_t adc_demo_remove(struct adc_demo_desc *desc)
 
+int32_t adc_get_number_of_channels(void* dev)
+{
+	struct adc_demo_desc *desc = dev;
```
check for 
```
if (!dev)
       return -ENODEV;
```
Do this for all other functions

mirceacaprioru replied:
Not sure where this function will be called do you check there that dev is not NULL ?  If you do this before calling this function then you don't need it here.

ranechita replied:
updated

mirceacaprioru commented on drivers/adc/adc_demo/adc_demo.c:
```diff
@@ -42,10 +42,33 @@
 /******************************************************************************/
 
 #include <stdlib.h>
+#include <math.h>
```
commit description should be a maximum of 75 characters/line

ranechita commented:
v4

-  added `if (!dev)
       return -ENODEV;` to functions that needed
- checked for '{' to be on the same line with 'if'/'for' etc

mirceacaprioru commented on drivers/adc/adc_demo/adc_demo.c:
```diff
@@ -85,13 +112,141 @@ int32_t adc_demo_remove(struct adc_demo_desc *desc)
+	if(!dev)
+		return -ENODEV;
+
+	struct adc_demo_desc *desc = dev;
```
leave the variable definition at function start just don't do the attribution operation until the check for !dev

ranechita commented:
Combined the changes from both branches into demo_driver, (draft) pull request can be found at https://github.com/analogdevicesinc/no-OS/pull/934
